File: app/src/main/java/com/hkb48/keepdo/calendar/CalendarFragment.java
package com.hkb48.keepdo.calendar;

import android.content.Context;
import android.content.Intent;
import android.graphics.Bitmap;
import android.graphics.Canvas;
import android.net.Uri;
import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.core.content.FileProvider;
import androidx.fragment.app.Fragment;
import androidx.viewpager2.adapter.FragmentStateAdapter;
import androidx.viewpager2.widget.ViewPager2;

import com.google.android.material.tabs.TabLayout;
import com.google.android.material.tabs.TabLayoutMediator;
import com.hkb48.keepdo.DatabaseAdapter;
import com.hkb48.keepdo.DateChangeTimeUtil;
import com.hkb48.keepdo.R;
import com.hkb48.keepdo.TaskDetailActivity;
import com.hkb48.keepdo.settings.Settings;
import com.hkb48.keepdo.util.CompatUtil;

import java.io.File;
import java.io.FileOutputStream;
import java.io.IOException;
import java.text.SimpleDateFormat;
import java.util.ArrayList;
import java.util.Calendar;
import java.util.Date;
import java.util.Locale;

public class CalendarFragment extends Fragment {

    /**
     * The maximum number of months to be paging up (Past: 10 years, Future: 1 year)
     */
    private static final int NUM_MAXIMUM_MONTHS_PAST = 10 * 12;  // 10 years
    static final int INDEX_OF_THIS_MONTH = NUM_MAXIMUM_MONTHS_PAST - 1;
    private static final int NUM_MAXIMUM_MONTHS_FUTURE = 12;  // 12 months
    private static final int NUM_MAXIMUM_MONTHS = NUM_MAXIMUM_MONTHS_PAST + NUM_MAXIMUM_MONTHS_FUTURE;
    private static final String FILE_PROVIDER = "com.hkb48.keepdo.fileprovider";

    private ViewPager2 mViewPager;

    public CalendarFragment() {
    }

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setHasOptionsMenu(true);
    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        return inflater.inflate(R.layout.fragment_calendar, container, false);
    }

    @Override
    public void onViewCreated(@NonNull View view, Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);

        mViewPager = view.findViewById(R.id.viewPager);
        mViewPager.setAdapter(new CalendarPageAdapter(this));
        mViewPager.setCurrentItem(CalendarFragment.INDEX_OF_THIS_MONTH, false);
        TabLayout tabLayout = view.findViewById(R.id.tab_layout);
        new TabLayoutMediator(tabLayout, mViewPager,
                (tab, position) -> tab.setText(getPageTitle(position))
        ).attach();
    }

    @Override
    public void onCreateOptionsMenu(@NonNull Menu menu, MenuInflater inflater) {
        inflater.inflate(R.menu.activity_task, menu);
        super.onCreateOptionsMenu(menu, inflater);
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        final long taskId = requireActivity().getIntent().getLongExtra("TASK-ID", -1);
        if (item.getItemId() == R.id.menu_info) {
            Intent intent = new Intent(requireActivity(), TaskDetailActivity.class);
            intent.putExtra("TASK-ID", taskId);
            startActivity(intent);
            return true;
        } else if (item.getItemId() == R.id.menu_share) {
            shareDisplayedCalendarView(taskId);
            return true;
        } else {
            return super.onOptionsItemSelected(item);
        }
    }

    private CharSequence getPageTitle(int position) {
        final int pageNumber = position - INDEX_OF_THIS_MONTH;

        Calendar current = DateChangeTimeUtil.getDateTimeCalendar();
        current.add(Calendar.MONTH, pageNumber);
        current.set(Calendar.DAY_OF_MONTH, 1);
        SimpleDateFormat sdf = new SimpleDateFormat("yyyy/MM", Locale.JAPAN);
        return sdf.format(current.getTime());
    }

    private void shareDisplayedCalendarView(final long taskId) {
        final String BITMAP_DIR_PATH = DatabaseAdapter.backupDirPath();
        final String BITMAP_FILE_PATH = BITMAP_DIR_PATH + "/temp_share_image.png";

        Context context = requireContext();
        View calendarRoot = requireActivity().findViewById(R.id.calendar_root);
        calendarRoot.setDrawingCacheEnabled(true);

        File bitmapFile = new File(BITMAP_FILE_PATH);
        if (!new File(BITMAP_DIR_PATH).mkdir()) {
            Bitmap bitmap = Bitmap.createBitmap(calendarRoot.getDrawingCache());

            Bitmap baseBitmap = Bitmap.createBitmap(bitmap.getWidth(),
                    bitmap.getHeight(), Bitmap.Config.ARGB_4444);
            Canvas bmpCanvas = new Canvas(baseBitmap);
            bmpCanvas.drawColor(CompatUtil.getColor(context,
                    R.color.calendar_bg_fargment));
            bmpCanvas.drawBitmap(bitmap, 0, 0, null);

            FileOutputStream fos = null;
            try {
                fos = new FileOutputStream(bitmapFile, false);
                baseBitmap.compress(Bitmap.CompressFormat.PNG, 100, fos);
                fos.flush();
                fos.close();
                calendarRoot.setDrawingCacheEnabled(false);
            } catch (Exception e) {
                e.printStackTrace();
            } finally {
                try {
                    if (fos != null) {
                        fos.close();
                    }
                } catch (IOException e) {
                    e.printStackTrace();
                }
            }

            bitmap.recycle();
            baseBitmap.recycle();
            Intent intent = new Intent(Intent.ACTION_SEND);
            intent.addFlags(Intent.FLAG_GRANT_READ_URI_PERMISSION);
            intent.setType("image/png");
            Uri contentUri = FileProvider.getUriForFile(context, FILE_PROVIDER, bitmapFile);
            intent.putExtra(Intent.EXTRA_STREAM, contentUri);
            DatabaseAdapter dbAdapter = DatabaseAdapter.getInstance(context);
            int comboCount = dbAdapter.getComboCount(taskId);
            String taskName = dbAdapter.getTask(taskId).getName();
            String extraText = "";
            int monthOffset = mViewPager.getCurrentItem() - CalendarFragment.INDEX_OF_THIS_MONTH;
            if (monthOffset == 0 && comboCount > 1) {
                extraText += context.getString(R.string.share_combo, taskName, comboCount);
            } else {
                Calendar current = DateChangeTimeUtil.getDateTimeCalendar();
                current.add(Calendar.MONTH, monthOffset);
                current.set(Calendar.DAY_OF_MONTH, 1);
                ArrayList<Date> doneDateList = dbAdapter.getHistoryInMonth(taskId, current.getTime());
                extraText += context.getString(R.string.share_non_combo, taskName, doneDateList.size());
            }
            extraText += " " + context.getString(R.string.share_app_url);
            intent.putExtra(Intent.EXTRA_TEXT, extraText);
            startActivity(intent);
        }
    }

    public static class CalendarPageAdapter extends FragmentStateAdapter {
        CalendarPageAdapter(Fragment fragment) {
            super(fragment);
        }

        @Override
        public int getItemCount() {
            return Settings.getEnableFutureDate() ? CalendarFragment.NUM_MAXIMUM_MONTHS : CalendarFragment.NUM_MAXIMUM_MONTHS_PAST;
        }

        @Override
        @NonNull
        public Fragment createFragment(int position) {
            return CalendarGrid.newInstance(position);
        }
    }
}

File: app/src/main/java/com/hkb48/keepdo/calendar/CalendarFragment.kt
package com.hkb48.keepdo.calendar

import android.app.Activity
import android.content.Intent
import android.graphics.Bitmap
import android.graphics.Canvas
import android.graphics.Rect
import android.net.Uri
import android.os.Build
import android.os.Bundle
import android.os.Handler
import android.os.Looper
import android.view.*
import androidx.core.content.FileProvider
import androidx.fragment.app.Fragment
import androidx.fragment.app.viewModels
import androidx.lifecycle.lifecycleScope
import androidx.viewpager2.adapter.FragmentStateAdapter
import androidx.viewpager2.widget.ViewPager2
import com.google.android.material.tabs.TabLayout
import com.google.android.material.tabs.TabLayoutMediator
import com.hkb48.keepdo.DateChangeTimeUtil
import com.hkb48.keepdo.R
import com.hkb48.keepdo.TaskDetailActivity
import com.hkb48.keepdo.databinding.FragmentCalendarBinding
import com.hkb48.keepdo.db.entity.Task
import com.hkb48.keepdo.settings.Settings
import com.hkb48.keepdo.util.CompatUtil
import com.hkb48.keepdo.viewmodel.TaskViewModel
import com.hkb48.keepdo.viewmodel.TaskViewModelFactory
import kotlinx.coroutines.launch
import java.io.File
import java.io.FileOutputStream
import java.text.SimpleDateFormat
import java.util.*


class CalendarFragment : Fragment() {
    private lateinit var mViewPager: ViewPager2
    private val taskViewModel: TaskViewModel by viewModels {
        TaskViewModelFactory(requireActivity().application)
    }
    private var _binding: FragmentCalendarBinding? = null
    private val binding get() = _binding!!

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setHasOptionsMenu(true)
    }

    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View {
        _binding = FragmentCalendarBinding.inflate(inflater, container, false)
        return binding.root
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        mViewPager = binding.viewPager
        mViewPager.adapter = CalendarPageAdapter(this)
        mViewPager.setCurrentItem(INDEX_OF_THIS_MONTH, false)
        val tabLayout = binding.tabLayout
        TabLayoutMediator(
            tabLayout, mViewPager
        ) { tab: TabLayout.Tab, position: Int ->
            tab.text = getPageTitle(position)
        }.attach()
    }

    override fun onDestroyView() {
        super.onDestroyView()
        _binding = null
    }

    override fun onCreateOptionsMenu(menu: Menu, inflater: MenuInflater) {
        inflater.inflate(R.menu.activity_task, menu)
        super.onCreateOptionsMenu(menu, inflater)
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        val taskId = requireActivity().intent.getIntExtra(
            TaskCalendarActivity.EXTRA_TASK_ID, Task.INVALID_TASKID
        )
        return when (item.itemId) {
            R.id.menu_info -> {
                startActivity(Intent(requireContext(), TaskDetailActivity::class.java).apply {
                    putExtra(TaskDetailActivity.EXTRA_TASK_ID, taskId)
                })
                true
            }
            R.id.menu_share -> {
                shareDisplayedCalendarView(taskId)
                true
            }
            else -> {
                super.onOptionsItemSelected(item)
            }
        }
    }

    private fun getPageTitle(position: Int): CharSequence {
        val pageNumber = position - INDEX_OF_THIS_MONTH
        val current = DateChangeTimeUtil.dateTimeCalendar
        current.add(Calendar.MONTH, pageNumber)
        current[Calendar.DAY_OF_MONTH] = 1
        val sdf = SimpleDateFormat("yyyy/MM", Locale.JAPAN)
        return sdf.format(current.time)
    }

    private fun shareDisplayedCalendarView(taskId: Int) {
        val calendarRoot = binding.calendarRoot
        getBitmapFromView(calendarRoot, requireActivity(), callback = {
            var contentUri: Uri? = null
            try {
                val imageDir = File(requireContext().cacheDir, "images").apply {
                    if (exists().not()) {
                        mkdir()
                    }
                }
                val file = File(imageDir, "shared_image.png")
                val fos = FileOutputStream(file)
                it.compress(Bitmap.CompressFormat.PNG, 100, fos)
                fos.flush()
                fos.close()
                contentUri = FileProvider.getUriForFile(requireContext(), FILE_PROVIDER, file)
            } catch (e: java.lang.Exception) {
                e.printStackTrace()
            }

            viewLifecycleOwner.lifecycleScope.launch {
                val comboCount = taskViewModel.getComboCount(taskId)
                val taskName = taskViewModel.getTask(taskId)?.name
                var extraText = ""
                val monthOffset = mViewPager.currentItem - INDEX_OF_THIS_MONTH
                if (monthOffset == 0 && comboCount > 1) {
                    extraText += requireContext().getString(
                        R.string.share_combo,
                        taskName,
                        comboCount
                    )
                } else {
                    val current = DateChangeTimeUtil.dateTimeCalendar
                    current.add(Calendar.MONTH, monthOffset)
                    current[Calendar.DAY_OF_MONTH] = 1
                    val doneDateList = taskViewModel.getHistoryInMonth(
                        taskId, current[Calendar.YEAR], current[Calendar.MONTH]
                    )
                    extraText += requireContext().getString(
                        R.string.share_non_combo,
                        taskName,
                        doneDateList.size
                    )
                }
                extraText += " " + requireContext().getString(R.string.share_app_url)

                startActivity(
                    Intent(Intent.ACTION_SEND).apply {
                        addFlags(Intent.FLAG_GRANT_READ_URI_PERMISSION)
                        type = "image/png"
                        putExtra(Intent.EXTRA_STREAM, contentUri)
                        putExtra(Intent.EXTRA_TEXT, extraText)
                    }
                )
            }
        })
    }

    private fun getBitmapFromView(view: View, activity: Activity, callback: (Bitmap) -> Unit) {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            activity.window?.let { window ->
                val bitmap = Bitmap.createBitmap(view.width, view.height, Bitmap.Config.ARGB_8888)
                val locationOfViewInWindow = IntArray(2)
                view.getLocationInWindow(locationOfViewInWindow)
                try {
                    PixelCopy.request(
                        window,
                        Rect(
                            locationOfViewInWindow[0],
                            locationOfViewInWindow[1],
                            locationOfViewInWindow[0] + view.width,
                            locationOfViewInWindow[1] + view.height
                        ),
                        bitmap,
                        { copyResult ->
                            if (copyResult == PixelCopy.SUCCESS) {
                                callback(bitmap)
                            }
                            // possible to handle other result codes ...
                        },
                        Handler(Looper.getMainLooper())
                    )
                } catch (e: IllegalArgumentException) {
                    // PixelCopy may throw IllegalArgumentException, make sure to handle it
                    e.printStackTrace()
                }
            }
        } else {
            @Suppress("DEPRECATION")
            view.isDrawingCacheEnabled = true

            @Suppress("DEPRECATION")
            val bitmap = Bitmap.createBitmap(view.drawingCache)
            val baseBitmap = Bitmap.createBitmap(
                bitmap.width,
                bitmap.height, Bitmap.Config.ARGB_8888
            )
            Canvas(baseBitmap).apply {
                drawColor(
                    CompatUtil.getColor(
                        activity,
                        R.color.calendar_bg_fargment
                    )
                )
                drawBitmap(bitmap, 0f, 0f, null)
            }
            callback(baseBitmap)
            bitmap.recycle()
            baseBitmap.recycle()
            @Suppress("DEPRECATION")
            view.isDrawingCacheEnabled = false
        }
    }

    class CalendarPageAdapter internal constructor(fragment: Fragment?) : FragmentStateAdapter(
        fragment!!
    ) {
        override fun getItemCount(): Int {
            return if (Settings.enableFutureDate) NUM_MAXIMUM_MONTHS else NUM_MAXIMUM_MONTHS_PAST
        }

        override fun createFragment(position: Int): Fragment {
            return CalendarGrid.newInstance(position)
        }
    }

    companion object {
        /**
         * The maximum number of months to be paging up (Past: 10 years, Future: 1 year)
         */
        private const val NUM_MAXIMUM_MONTHS_PAST = 10 * 12 // 10 years
        const val INDEX_OF_THIS_MONTH = NUM_MAXIMUM_MONTHS_PAST - 1
        private const val NUM_MAXIMUM_MONTHS_FUTURE = 12 // 12 months
        private const val NUM_MAXIMUM_MONTHS = NUM_MAXIMUM_MONTHS_PAST + NUM_MAXIMUM_MONTHS_FUTURE
        private const val FILE_PROVIDER = "com.hkb48.keepdo.fileprovider"
    }
}